### PR TITLE
Unittest sample : Proposal for issue #4

### PR DESCRIPTION
**Pull request by vinklein**
This pull request port some GP doctests as Python unittest.
This a solution proposal for issue #4. 
A more complete but not fully working version can be found here [unittest_dev](https://github.com/vinklein/cypari2/tree/unittest_dev).

What is your advices on this implementation ? 
Comments and critics are welcome

### CODE REVIEW

**Comment by videlec** [comment]
What is the file `unittest_sample` used for?

File: tests/unittests/alltest.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    'zeta'
+    ]
+
+require_galpol = ["galpol"]
+require_seadata = ["ellmodulareqn"]
+
+galpol_installed = True
+seadata_installed = True
+
+# test extensions presence.
```
Comment (videlec) [inline tests/unittests/alltest.py]: I would propose to encapsulate these in functions `has_seadata`, `has_galpol`, `has_elldata`, etc that would actually be part of the cypari2 library. Perhaps as methods of `Pari`.

**Comment by vinklein** [comment]
`unittest_sample` is this branch, if you mean `unittest_template`, it's a unused file i shouldn't have commit : )

File: tests/unittests/alltest.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    if "error opening galpol file" in str(e):
+        galpol_installed = False
+    else:
+        raise e
+
+suite = unittest.TestSuite()
+
+
+for t in testmodules:
+    if not ((t in require_galpol and not galpol_installed) or (t in require_seadata and not seadata_installed)):
```
Comment (videlec) [inline tests/unittests/alltest.py]: please change the condition to
```
(galpol_installed or t not in require_galpol) and (seadata_installed or t not in require_seadata)
```

File: tests/unittests/alltest.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        try:
```
Comment (videlec) [inline tests/unittests/alltest.py]: Couldn't this be uniform among all the files? This `try/except` is a bit ugly.

File: tests/unittests/alltest.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
+import unittest
```
Comment (videlec) [inline tests/unittests/alltest.py]: You should briefly document the mechanic of the tests here.

File: tests/unittests/bern.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+"""
+import unittest
+from cypari2 import Pari, PariError
+
+pari = Pari()
+
+
+class TestBern(unittest.TestCase):
+    def test_bern(self):
+        pari.bernfrac(0);
```
Comment (videlec) [inline tests/unittests/bern.py]: what this is supposed to test? and why keeping `;`?

File: tests/unittests/bern.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        pari.bernfrac(1);
```
Comment (videlec) [inline tests/unittests/bern.py]: idem

**Comment by jdemeyer** [comment]
I think that we should not add all those file to cypari2, especially because we want to support multiple versions of PARI and the version that we add is just one version.

Instead we should just use `tests_generator.py` (where is that script anyway?) and test directly the output of that.

**Comment by videlec** [comment]
I don't think that `test_generator.py` is actually generating the test files in their current form. This message is very misleading. Though it is important to keep the information from which PARI test it came from.

**Comment by vinklein** [comment]
The comments `# This file was generated by tests_generator.py` should be ignored for now (And removed).
test_generator.py exist but it only partly generate the unittest files, most of the work has been done manually. 
That's because the original gp doctest are written with a lot of different kind of syntax.

**Comment by jdemeyer** [comment]
Sorry for sounding a bit negative, but I don't quite understand why you are using unittests here. I thought that the main point of #4 was to have Python *doctests*. That is not what you doing here.

Now it seems that you just copy PARI tests to cypari2 which is a bit pointless because it's not the job of cypari2 to test PARI. It should be the job of cypari2 to test the wrapping mechanism, not to test the PARI library itself.

**Comment by vinklein** [comment]
> Sorry for sounding a bit negative, but I don't quite understand why you are using unittests here. I thought that the main point of #4 was to have Python doctests. That is not what you doing here.

An understanble question. If the main goal is to have Python doctest this PR have no sense. But the choice of unittests is more an answer to the question "What is the good way to port GP tests into cypari2".  
For the choice unittest vs doctest : 
- unittest is more flexible overall.
- IMO the main question when chosing betwwen unittest or doctest is to know if your tests are mainly here for informative value (showing what can be done) or if your tests are here to try to test every case.
GP tests looks more to be if the latter kind, and a lot of tests result are pretty unreadable given their size or structure (like testing many function with many parameters in nested loops) and have a poor value as documentation.

> It should be the job of cypari2 to test the wrapping mechanism, not to test the PARI library itself.

Can you expand on that ? 
Does test something like `pari.addprimes([pari.nextprime(1e9), pari.nextprime(1e10)])` has no added value for you ?

**Comment by videlec** [comment]
The GP tests are a natural source of code examples (that is moreover tested). This PR is not testing PARI/GP in any way but rather if GP code can easily be ported in cypari2. It checks for coherences and helps identifying what is the missing in cypari2 vs GP.

File: tests/unittests/bern.py (hunk 5 above)
Comment (vinklein) [reply tests/unittests/bern.py]: It test that it doesn't crash. I keep them mainly to stick with the original GP test file. 

I keep `;` during developement as an help to associate results with the corresponding test. 
Keeping them after can has no sense. I will remove them.

**Comment by jdemeyer** [comment]
> IMO the main question when chosing betwwen unittest or doctest is to know if your tests are mainly here for informative value (showing what can be done)

The GP doctests (i.e. the tests appearing in the PARI user documentation) are meant as documentation/examples, analogous to Python doctests.

The PARI unittests (i.e. the tests in the `src/pari` directory in the PARI sources) are meant as pure tests.

I think it makes most sense to port the GP doctests to cypari2. Porting the PARI unittests seems useless.

**Comment by vinklein** [comment]
> Porting the PARI unittests seems useless.

I  really don't see how it can be useless. Unless you have a way to ensure that cypari2 is perfectly iso-functional with pari. Saying it's useless it's the same as saying "the wrapping is perfect and will never fail". I thinks you mean it's not worth it.

Now what do we do with these unittest ?

**Comment by jdemeyer** [comment]
> the wrapping is perfect and will never fail

It's not that the wrapping will never fail. It's more like "the wrapping will either work or fail badly". It would be really surprising if the wrapping would work for *most* functions but not all.

Of course, the cases where we manually wrap some functions should be tested properly.

**Pull request closed without merging**